sequences/README.md

```diff
@@ -0,0 +1,257 @@
+
+Fine!
+
+What happens if we don't pass an array with two numbers, but `[12, "foo"]`?
+
+```sh
+$ curl -si -H "Content-Type: application/json" -d '[12, "foo"]' localhost:8080/connect
+HTTP/1.1 500 Internal Server Error
+Content-Type: application/json
+...
+
+{
+  "error": {
+    "id":      "c71kp4t916bht5pkuqag",
+    "message": "expression evaluation error",
+    "path":    "/multiply",
+    "status":  500
+  }
+}
+```
+
+Hmm, this is an error from the `/multiply` endpoint. But looking at the logs, we see that already the `/add` endpoint logged an error:
+
+```
+Invalid operand; Unsuitable value for right operand: a number is required.  ... endpoint=/add error_type=evaluation
+```
+
+An the `/multiply` endpoint logged another message:
+
+```
+Operation failed; Error during operation: argument must not be null.  ... endpoint=/multiply error_type=evaluation
+```
+
+Additionally, we see two entries in the `couper_backend` log, one for `/add` and one for `/multiply`. So Couper sent the second request, even though the first produced an unexpected result.
+
+But we can stop the sequence earlier by configuring the expected status code for each request:
+
+```hcl
+...
+      proxy "p" {
+        ...
+        expected_status = [200]                  # ←
+      }
+      # "default" request: pass response to client
+      request {
+        ...
+        expected_status = [200]                  # ←
+      }
+...
+```
+
+```sh
+$ curl -si -H "Content-Type: application/json" -d '[12, "foo"]' localhost:8080/connect
+HTTP/1.1 502 Bad Gateway
+Content-Type: application/json
+Couper-Error: endpoint error
+...
+
+{
+  "error": {
+    "id":      "c71kuad916bht5pkuqb0",
+    "message": "endpoint error",
+    "path":    "/connect",
+    "status":  502
+  }
+}
+```
+
+Now we see only one entry in the `couper_backend` log (for `/add`).
+
+We also see an error in the `couper_access` log:
+
+```
+endpoint error: endpoint error ... endpoint=/connect error_type=sequence
+```
+
+Let's add an `error_handler` to handle the error:
+
+```hcl
+...
+      request {
+        ...
+        expected_status = [200]
+      }
+      error_handler "unexpected_status" {
+        response {
+          status = 500
+          json_body = {
+            error = "upstream error"
+            error_description = "an upstream service responded with an unexpected status code"
+          }
+        }
+      }
+...
+```
+
+```sh
+$ curl -si -H "Content-Type: application/json" -d '[12, "foo"]' localhost:8080/connect
+HTTP/1.1 500 Internal Server Error
+Content-Type: application/json
+...
+
+{"error":"upstream error","error_description":"an upstream service responded with an unexpected status code"}
+```
+
+And we can log some additionaly information about the requests in the case of an error:
+
+```hcl
+...
+      error_handler "unexpected_status" {
+        response {
+          ...
+        }
+        custom_log_fields = {
+          p = backend_responses.p.json_body
```

Comment:
If the log format is json, logging `json_body` is easier to read:
```json
  "custom": {
    "add": {
      "result": 46
    },
    "default": {
      "error": {
        "id": "c7bvc3gplt7t9hpdkasg",
        "message": "expression evaluation error",
        "path": "/multiply",
        "status": 500
      }
    }
  }
```
instead of
```json
  "custom": {
    "add": "{\"result\":46}",
    "default": "{\n  \"error\": {\n    \"id\":      \"c7bvbgoplt7t9hpdkarg\",\n    \"message\": \"expression evaluation error\",\n    \"path\":    \"/multiply\",\n    \"status\":  500\n  }\n}\n"
  }
```

Reply:
in general it's not a good idea write a complete json body (be it request or response) to the log. that can easily lead to an explosion of log fields, and also field type clashes, depending on the log server in use.

for the purpose of this example, we could simply log `…json_body.error.message` (and maybe path)?

Reply:
ok, `path` can be derived from the config `url`; but if the error ist `unexpected_status` the actual `status` should be logged:

```hcl
        custom_log_fields = {
          add = {
            message = backend_responses.add.json_body.error.message
            status = backend_responses.add.json_body.error.status
          }
          default = {
            message = backend_responses.default.json_body.error.message
            status = backend_responses.default.json_body.error.status
          }
        }
```
creating
```json
{
  "custom":{
    "add":{
      "message":"expression evaluation error",
      "status":500
    },
    "default":{}
  },
  "error_type":"unexpected_status"
}
```